hnn_core/network.py

```diff
@@ -1187,6 +1187,160 @@ def add_connection(self, src_gids, target_gids, loc, receptor,
+        target_gids : str | int | range | list of int | None
+            Identifer for targets of source cells. Passing str arguments
+            ('L2_pyramidal', 'L2_basket', 'L5_pyramidal', 'L5_basket') is
+            equivalent to passing a list of gids for the relvant cell type.
+            source - target connections are made in an all-to-all pattern.
+        loc : str | list of str | None
+            Location of synapse on target cell. Must be
+            'proximal', 'distal', or 'soma'. Note that inhibitory synapses
+            (receptor='gabaa' or 'gabab') of L2 pyramidal neurons are only
+            valid loc='soma'.
+        receptor : str | list of str | None
+            Synaptic receptor of connection. Must be one of:
+            'ampa', 'nmda', 'gabaa', or 'gabab'.
+
+        Returns
+        -------
+        conn_indices : list of int
+            List of indices corresponding to items in net.connectivity.
+            Connecion indices are added if any of the provided parameter
+            values are present in a connection.
+
+        Notes
+        -----
+        Passing a list of values to a single parameter corresponds to a
+        logical OR operation across indices. For example,
+        loc=['distal', 'proximal'] returns all connections that target
+        distal or proximal dendrites.
+
+        Passing  multiple parameters corresponds to a logical AND operation.
+        For example, net.pick_connection(loc='distal', receptor='ampa')
+        returns only the indices of connections that target the distal
+        dendrites and have ampa receptors.
+        """
+
+        _validate_type(src_gids, (int, list, range, str, None), 'src_gids',
+                       'int list, range, str, or None')
+        _validate_type(target_gids, (int, list, range, str, None),
+                       'target_gids', 'int list, range or str')
+        _validate_type(loc, (str, list, None), 'loc', 'str, list, or None')
+        _validate_type(receptor, (str, list, None), 'receptor',
+                       'str, list, or None')
+
+        valid_srcs = list(self.gid_ranges.keys())  # includes drives as srcs
+        valid_targets = list(self.cell_types.keys())
+        valid_loc = ['proximal', 'distal', 'soma']
+        valid_receptor = ['ampa', 'nmda', 'gabaa', 'gabab']
+
+        # Convert src_gids to list
+        if src_gids is None:
+            src_gids = list()
+        elif isinstance(src_gids, int):
+            src_gids = [src_gids]
+        elif isinstance(src_gids, str):
+            _check_option('src_gids', src_gids, valid_srcs)
+            src_gids = self.gid_ranges[_long_name(src_gids)]
+        for src_gid in src_gids:
+            _validate_type(src_gid, int, 'src_gid')
+            gid_type = self.gid_to_type(src_gid)
+            if gid_type is None:
+                raise AssertionError(
+                    f'src_gid {src_gid} not in net.gid_ranges')
```

Comment:
can some of this be refactored with `add_connection`? I think it's time we created a file called `check.py` which has all these checks. See MNE and scikit-learn for inspiration:

https://github.com/mne-tools/mne-python/blob/main/mne/utils/check.py
https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/validation.py

Reply:
Created a `check,py` and moved a lot of tests there. In `add_connection` the `target_gids` still have some dedicated checks since they are turned into a more complicated structure, but in general it reduced a lot of redundancy.